File: src/Medium/TreeTest/Codec.java
package Medium.TreeTest;

public class Codec {


    public int i = 0;

    // Encodes a tree to a single string.
    // Encodes a tree to a single string.
    public String serialize(TreeNode root) {
        StringBuilder sb = new StringBuilder(  );
        toString(root ,sb);
        return sb.toString();
    }

    static void toString(TreeNode node,StringBuilder sb){
        if(node == null){
            return;
        }
        toString(node.left,sb);
        toString(node.right,sb);
        sb.append((char) node.val);
    }

    // Decodes your encoded data to tree.
    public TreeNode deserialize(String data) {
        if("".equals( data )){
            return null;
        }

        char[] datas =  data.toCharArray();
        i = datas.length -1;
        return toTree(datas,Integer.MIN_VALUE,Integer.MAX_VALUE);
    }

    TreeNode toTree( char[] datas, Integer min,Integer max){
        if(i < 0){
            return null;
        }
        char val = datas[i];
        if(val < min || val > max){
            return null;
        }
        i--;

        TreeNode node = new TreeNode(val);
        node.right = toTree(datas,node.val,max);
        node.left = toTree(datas,min,node.val);
        return node;
    }

    public static void main(String[] args) {
        StringBuilder builder=new StringBuilder();
        builder.append(10000);
        builder.append((char)520);
        System.out.println(builder.toString());
    }
}

File: src/Medium/TreeTest/Codec.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None
import json
from collections import deque
from math import inf

from src.Medium.TreeTest import TreeNode


class Codec:

    def serialize(self, root: TreeNode) -> str:
        """Encodes a tree to a single string.
        """

        if root is None:
            return " "
        return ','.join([str(root.val), self.serialize(root.left), self.serialize(root.right)])

    def deserialize(self, data: str) -> TreeNode:
        """Decodes your encoded data to tree.
        """

        queue = deque(data.split(","))

        def dfs():
            node = queue.popleft()
            if node == ' ':
                return None
            root = TreeNode(int(node))
            root.left = dfs()
            root.right = dfs()
            return root
        return dfs()

# Your Codec object will be instantiated and called as such:
# Your Codec object will be instantiated and called as such:
# ser = Codec()
# deser = Codec()
# tree = ser.serialize(root)
# ans = deser.deserialize(tree)
# return ans
